### ISSUE TITLE
[SF Migration] Allow injection of extra variables inside twig

### ISSUE THREAD

**Issue by matks** [PrestaShop/PrestaShop#13462]
Usecase: developers can override twig templates in order to modify a page display. For example if I prefer flat design over material design, I can re-style my BO with my design preference.

What happens if, while re-styling one template, I want to display an additional information using a variable that was not passed to Twig by the Controller ? As Symfony controllers are not overridable in PS 1.7, I might have to create a new Controller even though I just wanted to add 1 single extra variable. For example I just wanted to replace
```
return $this->render('@PrestaShop/Admin/Configure/AdvancedParameters/LogsPage/index.html.twig', [
            'layoutHeaderToolbarBtn' => [],
            'layoutTitle' => $this->trans('Logs', 'Admin.Navigation.Menu'),
            'requireAddonsSearch' => true,
            'grid' => $this->presentGrid($grid),
        ]);
```
by
```
return $this->render('@PrestaShop/Admin/Configure/AdvancedParameters/LogsPage/index.html.twig', [
            'layoutHeaderToolbarBtn' => [],
            'layoutTitle' => $this->trans('Logs', 'Admin.Navigation.Menu'),
            'requireAddonsSearch' => true,
            'grid' => $this->presentGrid($grid),
            'extra_var' => $myExtraVariable
        ]);
```

Possible solutions:
- Declare a twig global variable, but this impacts all templates, not just the one I wanted (suggestion by @mickaelandrieu)
- Create a dedicated Twig Extension to be used in my template (suggestion by @mickaelandrieu)
- Add a listener to KernelEVENTS::CONTROLLER and KernelEvents::VIEW and perform a hook call there to enable variables injection

**Comment by mickaelandrieu** [comment]
Good idea @matks 

=> https://stackoverflow.com/a/20353103

it sums up all possible solutions :-)

**Comment by matks** [comment]
An idea: introduce hook `actionBackOfficeControllerSetVariables()` ?

**Comment by tdavidsonas88** [comment]
@matks Do I understand correctly what a hook in this context means PS module hook?: https://devdocs.prestashop.com/1.7/modules/concepts/hooks/

**Comment by tdavidsonas88** [comment]
Found where to add the listener: https://github.com/PrestaShop/PrestaShop/blob/6d8ef75a43d737acf4a1846375556f187c263a9a/src/PrestaShopBundle/EventListener/ActionDispatcherLegacyHooksSubscriber.php#L62-L72

Not sure what the numbers: 100, 255 means

**Comment by tdavidsonas88** [comment]
Yeah, found the numbers meaning: _If different event subscriber methods listen to the same event, their order is defined by the priority parameter. This value is a positive or negative integer which defaults to 0. The higher the number, the earlier the method is called. Priority is aggregated for all listeners and subscribers, so your methods could be executed before or after the methods defined in other listeners and subscribers._   https://symfony.com/doc/3.4/event_dispatcher.html

**Comment by tdavidsonas88** [comment]
@matks I understand why to add listener to existing KernelEvents::CONTROLLER event - I can pass variables from Controller to twig. But why to also add it to KernelEvents::VIEW one more time?

**Comment by matks** [comment]
> @matks I understand why to add listener to existing KernelEvents::CONTROLLER event - I can pass variables from Controller to twig. But why to also add it to KernelEvents::VIEW one more time?

I did not know about callActionDispatcherBeforeHook and callActionDispatcherAfterHook so indeed maybe these 2 hooks are already enough to handle the topic of this issue.

Can we check that ? Because these hooks seem able to dispatch data, but do they allow to add extra variables into twig ? I think this would require to pass `event` to hook, right ?

**Comment by tdavidsonas88** [comment]
@matks We can add our new hook inside `public function callActionDispatcherBeforeHook(FilterControllerEvent $event)` and this would allow us to add new variable to the Request object of the rendered Controller. As I understood the task it is not enough to add our extraVar to the rendered Controller because then we would have check for that variable in every Controller; So I feel near of finding how to inject it directly to the controllers twig...

**Comment by matks** [comment]
What about extending Symfony Controller `render()` function to add our variables ?

```
class PrestaShopController extends Controller
{
    protected function render($view, array $parameters = [], Response $response = null)
    {
       $extraVars = $hookDispatcher->...();
       $parameters = array_merge($extraVars, $parameters);

       parent::render($view, $parameters, Response $response);
    }
}
```

**Comment by tdavidsonas88** [comment]
It might work, found something very similar: https://github.com/PrestaShop/PrestaShop/blob/2eb58e0c7c508d466e9a4d2e21e68e63a76b0ea8/src/PrestaShopBundle/Controller/Admin/FrameworkBundleAdminController.php#L47-L51

**Comment by tdavidsonas88** [comment]
@matks also hooks throws exception if I try to return array - so string is only supported

**Comment by tdavidsonas88** [comment]
There several things, that I search to do better like:
1. Parse from the module returned array/variable slightly better. Currently variable name is by default the same as module name; User defined variable would be better.

**Comment by tdavidsonas88** [comment]
> @matks also hooks throws exception if I try to return array - so string is only supported

Actually using the _dispatchRenderingWithParameters_  is capable to return a array: 
dashproducts -> array (
  'exVar1' => 'testvar1',
)
Just it's key is a module name by default

**Comment by matks** [comment]
PR started but not finished https://github.com/PrestaShop/PrestaShop/pull/17498

**Comment by comxd** [comment]
Since the PR https://github.com/PrestaShop/PrestaShop/pull/17498 is abandoned, what is the best way (without overriding a controller) to inject Twig vars into an override twig template?

**Comment by matks** [comment]
@comxd Maybe it would be a good try to rely on Symfony built-in kernel events https://symfony.com/doc/current/reference/events.html

The `kernel.view` or `kernel.controller` should allow you to insert code "just before the view is built" and inject the vars you need.

Not sure it will work but that's how I would try

**Comment by tom-combet** [comment]
I confirm that first solution works with `kernel.request` event. You can add a condition on the route to avoid passing variable to every controller/action like so:

```php
    public function onKernelRequest(GetResponseEvent $event)
    {
        $route = $event->getRequest()->attributes->get('_route');
        if ($route !== 'admin_orders_view') {
            return;
        }

        $myVar = 'test';

        $this->twig->addGlobal('myvar', $myVar);
    }
```
